public/app/features/explore/utils/decorators.ts

```diff
@@ -1,214 +1,187 @@
 import { config } from '@grafana/runtime';
+import { groupBy } from 'lodash';
 
-import { ExploreItemState, ExplorePanelData } from '../../../types';
+import { ExplorePanelData } from '../../../types';
 import { getGraphSeriesModel } from '../../../plugins/panel/graph2/getGraphSeriesModel';
 import { dataFrameToLogsModel } from '../../../core/logs_model';
 import { refreshIntervalToSortOrder } from '../../../core/utils/explore';
 
-export const decorateWithGraphLogsTraceAndTable = (
-  datasourceInstance?: DataSourceApi | null
-): OperatorFunction<PanelData, ExplorePanelData> => inputStream =>
-  inputStream.pipe(
-    map(data => {
-      if (data.error) {
-        return {
-          ...data,
-          graphFrames: [],
-          tableFrames: [],
-          logsFrames: [],
-          traceFrames: [],
-          graphResult: null,
-          tableResult: null,
-          logsResult: null,
-        };
-      }
-
-      const graphFrames: DataFrame[] = [];
-      const tableFrames: DataFrame[] = [];
-      const logsFrames: DataFrame[] = [];
-      const traceFrames: DataFrame[] = [];
+/**
+ * When processing response first we try to determine what kind of dataframes we got as one query can return multiple
+ * dataFrames with different type of data. This is later used for type specific processing. As we use this in
+ * Observable pipeline, it decorates the existing panelData to pass the results to later processing stages.
+ */
+export const decorateWithGraphLogsTraceAndTable = (data: PanelData): ExplorePanelData => {
+  if (data.error) {
+    return {
+      ...data,
+      graphFrames: [],
+      tableFrames: [],
+      logsFrames: [],
+      traceFrames: [],
+      graphResult: null,
+      tableResult: null,
+      logsResult: null,
+    };
+  }
 
-      for (const frame of data.series) {
-        if (shouldShowInVisualisationTypeStrict(frame, 'logs')) {
-          logsFrames.push(frame);
-        } else if (shouldShowInVisualisationTypeStrict(frame, 'graph')) {
+  const graphFrames: DataFrame[] = [];
+  const tableFrames: DataFrame[] = [];
+  const logsFrames: DataFrame[] = [];
+  const traceFrames: DataFrame[] = [];
+
+  for (const frame of data.series) {
+    switch (frame.meta?.preferredVisualisationType) {
+      case 'logs':
+        logsFrames.push(frame);
+        break;
+      case 'graph':
+        graphFrames.push(frame);
+        break;
+      case 'trace':
+        traceFrames.push(frame);
+        break;
+      case 'table':
+        tableFrames.push(frame);
+        break;
+      default:
+        if (isTimeSeries(frame)) {
           graphFrames.push(frame);
-        } else if (shouldShowInVisualisationTypeStrict(frame, 'trace')) {
-          traceFrames.push(frame);
-        } else if (shouldShowInVisualisationTypeStrict(frame, 'table')) {
           tableFrames.push(frame);
-        } else if (isTimeSeries(frame, datasourceInstance?.meta.id)) {
-          if (shouldShowInVisualisationType(frame, 'graph')) {
```

Comment:
This could have been removed because if it had any visualisation type meta info we would already match it before. So here we are only in case it does not have any preferences for vis type. This means we did not need the matcher functions anymore.